testonly/integration/clientserver.go

```diff
@@ -115,62 +115,95 @@ func getTestDB(testID string) (*sql.DB, error) {
 // testID should be unique to each unittest package so as to allow parallel tests.
-func NewLogEnv(testID string) (*LogEnv, error) {
+func NewLogEnv(ctx context.Context, numSequencers int, testID string) (*LogEnv, error) {
 	db, err := getTestDB(testID)
 	if err != nil {
 		return nil, err
 	}
 
 	timesource := &util.SystemTimeSource{}
-	registry, err := builtin.NewDefaultExtensionRegistry()
+	registry, err := builtin.NewExtensionRegistry(db)
 	if err != nil {
 		return nil, err
 	}
 
-	// Start Log Server.
+	// Create Log Server.
 	grpcServer := grpc.NewServer()
 	logServer := server.NewTrillianLogRPCServer(registry, timesource)
 	trillian.RegisterTrillianLogServer(grpcServer, logServer)
 
-	// Start Sequencer.
+	// Create Sequencer.
 	keyManager, err := crypto.LoadPasswordProtectedPrivateKey(privateKeyFile, privateKeyPassword)
 	if err != nil {
 		return nil, err
 	}
-	sequencerManager := server.NewSequencerManager(keyManager, registry,
-		sequencerWindow)
-	sequencerTask := server.NewLogOperationManagerForTest(ctx, registry,
-		batchSize, sleep, timeSource, sequencerManager)
+	sequencerManager := server.NewSequencerManager(keyManager, registry, sequencerWindow)
+	var wg sync.WaitGroup
+	var sequencerTask *server.LogOperationManager
+	var cancel context.CancelFunc
+	if numSequencers == 0 {
+		// Test sequencer that needs manual triggering (with env.Sequencer.OperationLoop()).
+		sequencerTask = server.NewLogOperationManagerForTest(ctx, registry,
+			batchSize, sleepBetweenRuns, timeSource, sequencerManager)
+	} else {
+		// Start a live sequencer in a goroutine.
+		var ctx2 context.Context
+		ctx2, cancel = context.WithCancel(ctx)
+		sequencerTask = server.NewLogOperationManager(ctx2, registry,
+			batchSize, numSequencers, sleepBetweenRuns, timesource, sequencerManager)
+		wg.Add(1)
+		go func(wg *sync.WaitGroup, om *server.LogOperationManager) {
+			defer wg.Done()
+			om.OperationLoop()
+		}(&wg, sequencerTask)
+	}
 
 	// Listen and start server.
 	addr, lis, err := listen()
 	if err != nil {
+		if cancel != nil {
+			cancel()
+		}
 		return nil, err
 	}
-	go grpcServer.Serve(lis)
+	wg.Add(1)
+	go func(wg *sync.WaitGroup, grpcServer *grpc.Server, lis net.Listener) {
+		defer wg.Done()
+		grpcServer.Serve(lis)
+	}(&wg, grpcServer, lis)
 
 	// Connect to the server.
 	cc, err := grpc.Dial(addr, grpc.WithInsecure())
 	if err != nil {
+		if cancel != nil {
+			cancel()
+		}
 		return nil, err
 	}
-
 	return &LogEnv{
-		grpcServer:   grpcServer,
-		logServer:    logServer,
-		ClientConn:   cc,
-		DB:           db,
-		LogOperation: sequencerManager,
-		Sequencer:    sequencerTask,
+		pendingTasks:    &wg,
+		grpcServer:      grpcServer,
+		logServer:       logServer,
+		ClientConn:      cc,
+		DB:              db,
+		LogOperation:    sequencerManager,
+		Sequencer:       sequencerTask,
+		sequencerCancel: cancel,
 	}, nil
 }
 
 // Close shuts down the server.
 func (env *LogEnv) Close() {
+	if env.sequencerCancel != nil {
```

Comment:
Consider replacing with `env.Sequencer.GracefulStop()` which should block until the sequencer has been stopped.